### PR TITLE
Dependency Injection -> RC2

### PR DESCRIPTION
**Pull request by ardalis**

### CODE REVIEW

File: aspnet/fundamentals/dependency-injection.rst [hunk 1]
```diff
@@ -31,18 +31,17 @@ ASP.NET Core includes a simple built-in container (represented by the ``IService
...
+  :lines: 37-50
   :dedent: 8
-  :emphasize-lines: 4,9,13
+  :emphasize-lines: 4,9,14
 
-The features and middleware provided by ASP.NET, such as MVC, follow a convention of using a single ``AddService()`` extension method to register all of the services required by that feature. 
+The features and middleware provided by ASP.NET, such as MVC, follow a convention of using a single Add\ *Service*\  extension method to register all of the services required by that feature. 
 
-.. note:: You can request certain framework-provided services within ``Startup`` methods - see :doc:`startup` for more details.
+.. tip:: You can request certain framework-provided services within ``Startup`` methods via their parameter lists - see :doc:`startup` for more details.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: via -> through
I think through is ESL/Translation friendly

File: aspnet/fundamentals/dependency-injection.rst [hunk 2]
```diff
@@ -51,51 +50,47 @@ Registering Your Own Services
 
 In the default web template example above, two application services are added to the ``IServiceCollection``.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: remove "default web template " as you did above, to keep it consistent.

File: aspnet/fundamentals/dependency-injection.rst [hunk 3]
```diff
@@ -31,18 +31,17 @@ ASP.NET Core includes a simple built-in container (represented by the ``IService
...
+The ``ConfigureServices`` method in the ``Startup`` class is responsible for defining the services the application will use, including platform features like Entity Framework Core and ASP.NET Core MVC. Initially, the ``IServiceCollection`` provided to ``ConfigureServices`` has just a handful of services defined. Below is an example of how to add additional services to the container using a number of extensions methods like ``AddDbContext``, ``AddIdentity``, and ``AddMvc``.
 
-.. literalinclude:: ../../common/samples/WebApplication1/src/WebApplication1/Startup.cs
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
   :language: c#
-  :linenos:
-  :lines: 39-56
+  :lines: 37-50
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: Your missing a much of lines like 
```
services.AddTransient<IEmailSender, AuthMessageSender>();
    services.AddTransient<ISmsSender, AuthMessageSender>();
```
which you refer to below

File: aspnet/fundamentals/dependency-injection.rst [hunk 4]
```diff
@@ -51,51 +50,47 @@ Registering Your Own Services
...
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
   :language: c#
-  :linenos:
-  :lines: 53-55
+  :lines: 52-54
   :dedent: 12
-  :emphasize-lines: 2-3
 
-.. note:: Keep in mind, in addition to these two types, everything else in ``ConfigureServices`` is about adding services. For example, ``services.AddMvc()`` adds the services MVC requires.
+.. note:: Keep in mind, in addition to these two types, everything else in ``ConfigureServices`` is also adding and configuring services. For example, ``services.AddMvc()`` adds the services MVC requires.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: I don't think you need this note. IF you want to keep it, lose all the dead wood.
Each ``services.Add<service>`` call adds (and potentially configures) services .For example, ``services.AddMvc`` add MVC.

File: aspnet/fundamentals/dependency-injection.rst [hunk 5]
```diff
@@ -51,51 +50,47 @@ Registering Your Own Services
...
 
 .. note:: Creating the requested object, and all of the objects it requires, and all of the objects those require, is sometimes referred to as an `object graph`. Likewise, the collective set of dependencies that must be resolved is typically referred to as a `dependency tree` or `dependency graph`.
 
-In this case, both ``ICharacterRepository`` and in turn ``ApplicationDbContext`` must be registered with the services container in ``ConfigureServices`` in ``Startup``. ``ApplicationDbContext`` is configured via the call to the extension method ``AddEntityFramework`` which includes an extension for adding a ``DbContext`` (``AddDbContext<T>``). Registration of the repository is done at the bottom end of ``ConfigureServices``:
+In this case, both ``ICharacterRepository`` and in turn ``ApplicationDbContext`` must be registered with the services container in ``ConfigureServices`` in ``Startup``. ``ApplicationDbContext`` is configured via the call to the extension method ``AddEntityFramework`` which includes an extension for adding a ``DbContext`` (``AddDbContext<T>``). Registration of the repository is done in ``ConfigureServices``:
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: Replace  Registration of the repository is done in ``ConfigureServices``: 
with 
The following code shows the CharacterRepository registration (or something like that)
Then use:

.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
  :language: c#
  :lines: 37-55,63
  :dedent: 8
  :emphasize-lines: 19

File: aspnet/fundamentals/dependency-injection.rst [hunk 6]
```diff
@@ -51,51 +50,47 @@ Registering Your Own Services
...
 
 .. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
   :language: c#
-  :lines: 62-63
-  :linenos:
+  :lines: 55
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: .. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
  :language: c#
  :lines: 37-55,63
  :dedent: 8
  :emphasize-lines: 19

File: aspnet/fundamentals/dependency-injection.rst [hunk 7]
```diff
@@ -113,35 +108,29 @@ Scoped
...
 
 Services can be registered with the container in several ways. We have already seen how to register a service implementation with a given type by specifying the concrete type to use. In addition, a factory can be specified, which will then be used to create the instance on demand. The third approach is to directly specify the instance of the type to use, in which case the container will never attempt to create an instance.
 
 To demonstrate the difference between these four lifetime and registration options, consider a simple interface that represents one or more tasks as an *operation* with a unique identifier, ``OperationId``. Depending on how we configure the lifetime for this service, the container will provide either the same or different instances of the service to the requesting class. To make it clear which lifetime is being requested, we will create one type per lifetime option:
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: You no longer have four. So just make it 
To demonstrate the difference between these lifetimes, ...

File: aspnet/fundamentals/dependency-injection.rst [hunk 8]
```diff
@@ -113,35 +108,29 @@ Scoped
...
 
 .. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Interfaces/IOperation.cs
   :language: c#
-  :linenos:
-  :emphasize-lines: 7
+  :emphasize-lines: 5,7
 
 We also implement all of these interfaces using a single class, ``Operation``, that simply accepts a ``Guid`` in its constructor, or uses a new ``Guid`` if none is provided.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: drop all

File: aspnet/fundamentals/dependency-injection/sample/src/DependencyInjectionSample/Services/OperationService.cs [hunk 9]
```diff
@@ -7,17 +7,17 @@ public class OperationService
         public IOperationTransient TransientOperation { get; private set; }
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection/sample/src/DependencyInjectionSample/Services/OperationService.cs]: use more of C#6 and get rid of unnecessary private set;

File: aspnet/fundamentals/dependency-injection.rst [hunk 10]
```diff
@@ -158,21 +147,20 @@ Observe which of the ``OperationId`` values varies within a request, and between
...
+Request Services represent the services you configure and request as part of your application. When your objects specify dependencies, these are satisfied by the types found in ``RequestServices``, not ``ApplicationServices``.
 
 Generally, you shouldn't use these properties directly, preferring instead to request the types your classes you require via your class's constructor, and letting the framework inject these dependencies. This yields classes that are easier to test (see :doc:`/testing/index`) and are more loosely coupled.
 
-.. note:: The important things to remember are that your application will almost always use ``RequestServices``, and in any case you shouldn't access these properties directly. Instead, request the services you need via your class's constructor. 
+.. warning:: Prefer requesting dependencies as constructor parameters to accessing the ``RequestServices`` collection.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: Nice rewrite, but wouldn't note be more appropriate?

File: aspnet/fundamentals/dependency-injection.rst [hunk 11]
```diff
@@ -232,7 +219,7 @@ Finally, configure Autofac as normal in ``DefaultModule``:
     }
   }
 
-Now at runtime, Autofac will be used to resolve types and inject dependencies.
+Now at runtime, Autofac will be used to resolve types and inject dependencies. `Learn more about using Autofac and ASP.NET Core <http://docs.autofac.org/en/latest/integration/aspnetcore.html>`_
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/dependency-injection.rst]: Drop now.

**Comment by Rick-Anderson** [comment]
Looks good to me with these fixed. @danroth27  do you want someone to take a quick look at ``IOperationSingletonInstance `` ?

File: aspnet/fundamentals/dependency-injection.rst (hunk 1 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: fixed

File: aspnet/fundamentals/dependency-injection.rst (hunk 2 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 3 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: Those lines are shown in the include on lines 53-56 of the rst file

File: aspnet/fundamentals/dependency-injection.rst (hunk 4 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 5 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 6 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 7 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 8 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: done

File: aspnet/fundamentals/dependency-injection/sample/src/DependencyInjectionSample/Services/OperationService.cs (hunk 9 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection/sample/src/DependencyInjectionSample/Services/OperationService.cs]: done

File: aspnet/fundamentals/dependency-injection.rst (hunk 10 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: Ok. I've gotten pretty strong guidance from the team about avoiding using RequestServices so was trying to be a bit more firm about it.

File: aspnet/fundamentals/dependency-injection.rst (hunk 11 above)
Comment (ardalis) [inline aspnet/fundamentals/dependency-injection.rst]: Done

**Comment by ardalis** [comment]
@Rick-Anderson @danroth27 what is it about IOperationSingleInstance you're concerned with? It's just showing how you can use Singleton with an instance specified at startup, as opposed to just letting the container instantiate the type the first time it is requested.

**Comment by Rick-Anderson** [comment]
:shipit:

**Comment by danroth27** [comment]
Merged with https://github.com/aspnet/Docs/pull/1232

**Pull request closed without merging**